source/language/delays.rst

```diff
@@ -287,47 +287,161 @@ to properly take into account the finite length of each gate.
+the other side of a box are permitted, but moving operations either into or out of the box as
+part of an optimization is forbidden. The compiler can also be given a description of the
+operation which a ``box`` definition is meant to realise, allowing it to re-order gates around
+the box. For example, consider a dynamical decoupling sequence inserted in a part of the circuit:
 
 .. code-block:: c
 
-       boxas mybox {
-           cx q[0], q[1];
-           delay[200ns] q[0];
-       }
-       delay[length(mybox)] q[2], q[3];
-       cx q[2], q[3];
+    rx(2*π/12) q;
+    box {
+        delay[ddt] q;
+        x q;
+        delay[ddt] q;
+        x q;
+        delay[ddt] q;
+    }
+    rx(3*π/12) q;
+
+By boxing the sequence, we create a box that implements the identity. The compiler is now free
+to commute a gate past the box by knowing the unitary implemented by the box:
+
+.. code-block:: c
+
+    rx(5*π/12) q;
+    box {
+        delay[ddt] q;
+        x q;
+        delay[ddt] q;
+        x q;
+        delay[ddt] q;
+    }
+
+The compiler can thus perform optimizations without interfering with the implmentation of the
+dynamical decoupling sequence. 
+
+As with other operations, we may use square brakets to assign a duration to a box: this can be
+used to put hard constraints on the execution of a particular sub-circuit by requiring it to
+have the assigned duration. This can be useful in scenarios where the exact duration of a piece
+of code is unknown (*e.g.*, if it is runtime dependent), but where it would be helpful to impose
+a duration on it for the purpose of scheduling the larger circuit. For example, if the duration
+of the parameterized gates ``mygate1(a, b), mygate2(a, b)`` depend on values of the variables
+``a`` and ``b`` in a complex way, but an offline calculation has shown that the total will never
+require more than 150ns for all valid combinations:
+
+.. code-block:: c
+
+    // some complicated circuit that gives runtime values to a, b
+    box [150ns] {
+        delay[str1] q1; // Schedule as late as possible within the box
+        mygate1(a, a+b) q[0], q[1];
+        mygate2(a, a-b) q[1], q[2];
+        mygate1(a-b, b) q[0], q[1];
+    }
+
+Boxes can be dressed by a ``verbatim`` modifier to prevent the compiler from optimizing the code
+piece inside.  The ``verbatim`` modifier is simply a compiler flag to let compiler passes ignore the
+boxed region. For example, gates will not be re-ordered or be cancelled in a ``verbatim`` box. The
+only two compiler passes that look inside a ``verbatim`` box are (1) the pass the resolves timing and
+duration (2) the pass that links gates to their ``defcal``. As a result, having any high-level constructs
+in a ``verbatim`` box that requires other compiler passes will result in a runtime error since related 
+compiler passes will not be able to look inside to properly compile them. Examples include:
```

Comment:
The `verbatim` proposal has not yet been accepted.

Reply:
Will remove. This PR is modified based on the `verbatim_box` PR.

Reply:
I will just open a new PR since the name of this PR becomes confusing.

Reply:
@blakejohnson Opened a new PR:https://github.com/Qiskit/openqasm/pull/247. Closing this one.